Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,189 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+              Submodule.injective_subtype E) ?_
+    rw [← hy hi, hr hi, smul_tmul, map_smul, LinearMap.rTensor_tmul, Submodule.subtype_apply, eq1,
+      smul_zero, map_zero]
+  have : Subsingleton (E ⊗[R] M) := subsingleton_iff_forall_eq 0 |>.2 fun x =>
+    show x ∈ (⊥ : Submodule R _) from eq0 ▸ ⟨⟩
+
+  -- but `E ⊗ M = 0` implies `E = 0` because `M` is faithfully flat and this is a contradiction.
+  exact not_subsingleton_iff_nontrivial.2 inferInstance <| fl.rTensor_reflects_triviality R M E
+
+lemma rTensor_reflects_exact [fl : FaithfullyFlat R M]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    Function.Exact l12 l23 := LinearMap.exact_iff.2 <| by
+  have complex : LinearMap.range l12 ≤ LinearMap.ker l23 := range_le_ker_of_exact_rTensor R M _ _ ex
+  -- by previous lemma, we need only to show `ker l23 ≤ range l12`.
+  refine le_antisymm ?_ complex
+  -- We consider the cohomology of `N1 -> N2 -> N3`, i.e. let `H := ker l23 ⧸ range l12`. It is
+  -- enough to show `H = 0` because `H = 0` implies `x ∈ range l12`.
+  let H := LinearMap.ker l23 ⧸ LinearMap.range (Submodule.inclusion complex)
+  suffices triv_coh : Subsingleton H by
+    rintro x (hx : l23 x = 0)
+    have eq0 : (Submodule.mkQ _ ⟨x, hx⟩ : H) = 0 := triv_coh.elim _ _
+    obtain ⟨⟨y, hy⟩, eq0⟩ := Submodule.Quotient.mk_eq_zero _ |>.1 eq0
+    simp only [Subtype.ext_iff, Submodule.coe_inclusion] at eq0
+    exact eq0 ▸ hy
+
+  -- Since `M` is faithfully flat, we need only to show that `H ⊗ M` is trivial.
+  suffices Subsingleton (H ⊗[R] M) from rTensor_reflects_triviality R M H
+  let e : H ⊗[R] M ≃ₗ[R] _ := TensorProduct.quotientTensorEquiv _ _
+  rw [e.toEquiv.subsingleton_congr, Submodule.subsingleton_quotient_iff_eq_top, eq_top_iff]
+  -- note that `H ⊗ M ≅ (ker l23 ⊗ M) ⧸ (range l12 ⊗ M)`, so we only need to show that
+  -- `ker l23 ⊗ M ≤ range l12 ⊗ M` as submodules of `ker l23 ⊗ M`.
+  set ι : (LinearMap.ker l23) ⊗[R] M →ₗ[R] N2 ⊗[R] M := (Submodule.subtype _).rTensor M with ι_def
+  -- Since `M` is flat and `ker l23 -> N2` is injective, we can compare them as submodules `N2 ⊗ M`.
+  rw [← Submodule.map_le_map_iff_of_injective (f := ι)
+    (hf := Module.Flat.rTensor_preserves_injective_linearMap _ Subtype.val_injective),
+    Submodule.map_top, show LinearMap.range ι = LinearMap.range (LinearMap.rTensor M l12) by
+      rw [← LinearMap.exact_iff.1 ex, ι_def]
+      ext x; constructor
+      · rintro ⟨x, rfl⟩
+        induction x using TensorProduct.induction_on <;> aesop
+      · rw [LinearMap.exact_iff.1 ex]
+        rintro ⟨x, rfl⟩
+        induction x using TensorProduct.induction_on with
+        | zero => simp
+        | tmul a b =>
+          exact ⟨⟨l12 a, complex (by simp)⟩ ⊗ₜ b, rfl⟩
+        | add x y hx hy => rw [map_add]; exact Submodule.add_mem _ hx hy,
+    LinearMap.rTensor, TensorProduct.map_range_eq_span_tmul, Submodule.span_le]
```

Comment:
I was following along just fine with the commented proof but now it just stopped?

Reply:
more comments is added